simpledotorg/simple-server
# Update "Hypertension Patients" card

**Story card:** [ch6168](https://app.shortcut.com/simpledotorg/story/6168/add-new-hypertension-coverage-progress-bar-and-data-to-hypertension-patients-card-in-state-and-district-reports)

## Because

The "Hypertension patients" card in the "Reports, Details" page will now display the new "Hypertension patient coverage" and "Total estimated hypertensive population" data.

**Note:** `Total registered patients = Patients under care + Lost to follow-up + Dead patients`
**Note:** `Hypertensive patient coverage = Total registered patients / Total estimated hypertensive population`

## This addresses

We need to update the following to show the new estimated hypertensive data:
1. Reorder the first two cards in Report "Details": (1) "Hypertension patients" (2) "Lost to follow-up"
2. Update the "Hypertension patients" card subtitle and remove the tooltip from the header
3. Show the new "Hypertension patient coverage" section in the "Hypertension patients" card for State and District reports
4. Show the new table layout in the "Hypertension patients" card for Block and Facility reports
5. Add new tooltips for each key indicator shown in the "Hypertension patients" card (Note: The "Total estimated hypertensive patients" definition is different for State and District reports)
6. Remove the "Transferred-out patients" definition from the "Details" footnotes
7. Add a "Total estimated hypertensive patients" definition to the "Details" footnotes (Note: The "Total estimated hypertensive patients" definition is different for State and District reports)

**State report**
![image](https://user-images.githubusercontent.com/16785131/146251530-475e147f-936e-4514-a00a-88ea740ed4e1.png)

**District report**
![image](https://user-images.githubusercontent.com/16785131/146251876-aae45d70-5ee4-41dc-ba58-5e6e99807c7d.png)

**Block report**
![image](https://user-images.githubusercontent.com/16785131/146251936-bbd8d065-73df-4c0b-a6e9-c184370d6ff0.png)

**Facility report**
![image](https://user-images.githubusercontent.com/16785131/146251970-180ed1e4-6b5f-41a5-a5a4-6913802f6375.png)

## Test instructions

1. In your local, `git checkout estimated-population-card` and `git pull`
2. Reinstall your gems with `bundle install`, update your database with `rails db:migrate`, and start your server `rails server`
3. Sign in with the `admin@simple.org` email. Click the "More, Flipper" menu option and add an `estimated_population` feature flag and enable it.
4. Go to the [reports home page](http://localhost:3000/reports/regions)
5. Open one of the states and make note of the names of the district regions. You'll need them to update their populations.
6. Next, go to the [facilities admin page](http://localhost:3000/admin/facilities)
7. Make sure each of the district regions doesn't have a `Total estimated hypertensive population` value (it's the subtitle beneath the district card's name). If a district does have a value, click the "Edit district" button and delete the value in the "Estimated hypertensive population" field. Save the changes.
8. Now, go back to the [reports home page](http://localhost:3000/reports/regions) and click the parent state of the districts you configured
9. Click the "Details" tab and make sure (1) `Hypertension patient coverage = N/A` (2) the progress bar is gray and the left and right sides have curved edges (3) `Total registered patients = Patients under care + Lost to follow-up + Dead patients` and (4) `Total est. hypertensive population = + Add district populations`
10. Scroll down to the footnotes and verify the `Hypertension coverage rate` and `Total estimated hypertensive population` definitions are displayed
11. Click the `+ Add district populations` link and make sure it takes you to the [facilities admin page](http://localhost:3000/admin/facilities)
12. Select one of the districts you took note of and enter an `Estimated hypertension population`
13. Visit that district's report details page, and make sure (1) `Hypertension patient coverage = (Total registered patients / Total est. hypertensive population) * 100` (2) the progress bar is showing a green progress bar (3) `Total registered patients = Patients under care + Lost to follow-up + Dead patients` and (4) `Total est. hypertensive population` is the value you entered
14. Scroll down to the footnotes and verify the `Hypertension coverage rate` and `Total estimated hypertensive population` definitions are displayed
15. Go back to the facilities admin page and enter a population value for the remaining districts
16. Go to the state's details report page and make sure  (1) `Hypertension patient coverage = (Total registered patients / Total est. hypertensive population) * 100` (2) the progress bar is showing a green progress bar (3) `Total registered patients = Patients under care + Lost to follow-up + Dead patients` and (4) `Total est. hypertensive population` is the sum of the values you entered for each district
17. Go to any facility details report page and make sure the `Hypertension patient coverage` section and progress bar aren't showing and that `Total registered patients = Patients under care + Lost to follow-up + Dead patients`
18. Go to any block details report page and make sure the `Hypertension patient coverage` section and progress bar aren't showing and that `Total registered patients = Patients under care + Lost to follow-up + Dead patients`
19. Open Flipper and deactivate the `estimated_population` feature flag
20. Go to a state, district, block, and facility report "Details" page and verify only the following key indicators are displayed: `Total registered patients`, `Patients under care`, `Lost to follow-up`, and `Dead patients`
21. Go to a state and district report "Details" page and make sure the `Hypertension coverage rate` and `Total estimated hypertensive population` definitions aren't displayed in the footnotes
22. Go to the "Facilities" admin page and verify each district card _does not_ have an estimated hypertensive population subtitle
23. Click the "Edit district" for one of the district cards and verify the `Estimated hypertensive population` field is not displayed.
